apps/prairielearn/elements/pl-rich-text-editor/pl-rich-text-editor.js

```diff
@@ -94,4 +94,38 @@
   MathFormula.tagName = 'SPAN';
 
   Quill.register('formats/formula', MathFormula, true);
+
+  class Counter {
+    constructor(quill, options) {
+      this.quill = quill;
+      this.options = options;
+      this.container = document.querySelector(options.container);
+      if (this.options.unit !== 'none') {
+        quill.on(Quill.events.TEXT_CHANGE, this.update.bind(this));
+      }
+    }
+
+    calculate() {
+      const text = this.quill.getText();
+
+      // FIXME: Support unicode characters in character count
+      if (this.options.unit === 'word') {
+        const trimmed = text.trim();
+        // Splitting empty text returns a non-empty array
+        return trimmed.length > 0 ? trimmed.split(/\s+/).length : 0;
+      } else if (this.options.unit === 'character') {
+        return text.length;
+      } else {
+        console.error(`Text count not implemented for unit type: ${this.options.unit}`);
+      }
+    }
+
+    update() {
+      const length = this.calculate();
+      const label = `${this.options.unit}${length === 1 ? '' : 's'}`;
+      this.container.innerText = `${length} ${label}`;
+    }
+  }
+
+  Quill.register('modules/counter', Counter);
```

Comment:
We already have, in this editor, a handler for the text-change event. I'm wondering if it may be simpler to just use that handler instead of creating a totally separate module for this functionality. This would make the code shorter and easier to maintain in the future.

Reply:
Keeping the components separate could help us easily tweak the priorities of different change events - for extremely large inputs, we could theoretically debounce character/word count updates but keep the inputElement setter as-is because we don't want to lose any unsaved changes. Of course, it is not impossible (maybe slightly more difficult) when the change events are merged.

Reply:
That is an interesting point, though I don't see how working with a module would make this easier to handle. AFAICT you can always have multiple handlers for the same event, if this is a concern of yours.

I see a module as an option when working with something that will be migrated across multiple instances of quill, including cases you don't have control over, which is not the case here. In this context a simple event handler is probably enough.

Reply:
If I was using rx, I'd create one subscription per on-change handler and apply different debouncings although now that I think about it, I can manually create two different event streams from one handler. Not a big difference syntactically imo.